core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -322,6 +324,19 @@ private Object processWithExceptionHandling(Expression expression, SymbolResolve
             }
         }
 
+        private Object processWithCaching(Expression expression, SymbolResolver resolver)
+        {
+            if (optimize && resolver instanceof NoOpSymbolResolver) {
+                // We are using weak reference map as cache, that's why we can't depend on an intermediate object
+                // that consists of expression as well as symbolResolver as a key. This is because intermediate object could
+                // get GCed even when expression still exists which will cause cache miss. So, to solve this problem we
+                // can only cache when context is noop. It'll also give us good performance benefits because in most of
+                // places optimize is called with noop symbol resolver.
+                return optimizationCache.computeIfAbsent(expression, key -> process(key, resolver));
+            }
+            return process(expression, resolver);
+        }
+
```

Comment:
nit:

This is sill missing improvement in `io.trino.sql.planner.ExpressionInterpreter.Visitor#visitInPredicate`

Within method there is `for (Expression expression : valueList.getValues()) {` loop which iterates over all elements. One of it's purposes is to optimize IN list terms.

You could cache optimization result for `valueList` so these elements are not optimized again. Otherwise if symbol is changed for expression like:
```
x IN (CAST('a' as VARCHAR(42), ...);
```
to
```
y IN (CAST('a' as VARCHAR(42), ...);
```
then with current code IN list will get reevaluated.